core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java

```diff
@@ -164,7 +167,8 @@ private void updateColumn(Types.NestedField field, Types.NestedField existingFie
 
     boolean needsOptionalUpdate = field.isOptional() && existingField.isRequired();
     boolean needsTypeUpdate =
-        field.type().isPrimitiveType() && !field.type().equals(existingField.type());
+        (field.type().isPrimitiveType() && !field.type().equals(existingField.type()))
```

Comment:
I think we may just want to extract all of this logic now into a helper function since we already have some of it in ignorableUpdate. We also only probably need to pass through the types rather than the whole field.